Mathlib/ModelTheory/Syntax.lean

```diff
@@ -759,10 +758,28 @@ noncomputable def iExs [Finite β] (φ : L.Formula (α ⊕ β)) : L.Formula α :
   let e := Classical.choice (Classical.choose_spec (Finite.exists_equiv_fin β))
   (BoundedFormula.relabel (fun a => Sum.map id e a) φ).exs
 
+variable (β) in
+/-- `iExsUnique f φ` transforms a `L.Formula (α ⊕ β)` into a `L.Formula β` by existentially
+quantifying over all variables `Sum.inr _` and asserting that the solution should be unique  -/
+noncomputable def iExsUnique [Finite β] (φ : L.Formula (α ⊕ β)) : L.Formula α :=
+  let _ := Fintype.ofFinite β
+  iExs β (φ ⊓ iAlls β
+    ((φ.relabel (fun a => Sum.elim (Sum.inl ∘ Sum.inl) Sum.inr a)).imp
+    (BoundedFormula.iInf
+      (fun g => Term.equal (var (Sum.inr g)) (var (Sum.inl (Sum.inr g)))))))
+
 /-- The biimplication between formulas, as a formula. -/
 protected nonrec abbrev iff (φ ψ : L.Formula α) : L.Formula α :=
   φ.iff ψ
 
+/-- take the disjunction of finitely many formulas -/
+noncomputable def iSup [Finite α] (f : α → L.Formula β) : L.Formula β :=
+  BoundedFormula.iSup f
+
+/-- take the conjunction of finitely many formulas -/
+noncomputable def iInf [Finite α] (f : α → L.Formula β) : L.Formula β :=
+  BoundedFormula.iInf f
```

Comment:
Do we need this? `Formula` is an `abbrev` for `BoundedFormula`, so `BoundedFormula.iSup`/`BoundedFormula.iInf` does everything you new definitions already do (and your new definitions aren't `abbrev`s)

Reply:
I don't know. Personally I think that the current `BoundedFormula` and `Formula` duplication is quite annoying but so far we do seem to have been duplicating a lot of things (but the library is not completely consistent). `BoundedFormula` can be a bit anooying to use when reasoning about semantics because for a `BoundedFormula L α 0` you have this term `Fin 0 -> M` at the end when you use `BoundedFormula.realize` which are always equal obviously, but not necessarily definitionally equal.

Reply:
The best long term solution in my view is probably to claim that `BoundedFormula` is just an implementation detail and we should basically always be using `Formula` and we just have to introduce a single variable `all` and `ex` on formulas and the exact type of these definitions I'm not sure of. I don't have a strong opinion on any temporary fix before that.

Reply:
I agree that there's some major cleanup that needs to happen in the long term - I wish I had a concrete suggestion for how it should go.

In the meantime, do you have a use case for `iSup` or `iInf` for `Formula`s as opposed to `BoundedFormula`s? Is it possible that we could even remove the `BoundedFormula` version and just work with `Formula`s now?

I'm personally fine having both if there's actual code that works better that way while awaiting a good general solution to duplication.

Reply:
I almost never used `BoundedFormula`, I think mainly because I was very rarely quantifying over a single variable, it was always over an arbitrary finite number of variables.

A use case for iInf is the statement of Ax Grothendieck. It's about injectivity and surjectivity of maps K^n -> K^n and I needed iInf to state equality of elements of K^n